Remove DimensionedTensorType

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24077 Remove DimensionedTensorType**

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.


This implementation keeps DimensionedTensorType because CompleteTensorType inherits from it. A second patch will remove both, and a third patch will merge ProfiledTensorType with TensorType.

Differential Revision: [D16729581](https://our.internmc.facebook.com/intern/diff/D16729581)

zdevito committed:
Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

zdevito committed:
Update on "Remove DimensionedTensorType"

Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

gh-metadata: pytorch pytorch 24077 gh/zdevito/81/head

Krovatkin commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -449,10 +449,11 @@ class ShapePropagator {
     }
     node->output()->setType(
-        DimensionedTensorType::create(default_type, default_device, dims));
+        ProfiledTensorType::create(default_type, default_device, dims, true));
```
nitpick: `/*requires_grad=*/ true`

apaszke replied:
Also, shouldn't all those places have requires_grad set to unknown?

zdevito replied:
Either unknown or true will end up doing the same thing. This was to preserve what dimensioned tensor did, which was always `true` by default. I can change it though.

apaszke replied:
Well unless we want to use the same overapproximation behavior as in DimensionedTensorType, then we should use unknown, because that's the right thing to o in this case. I thought the idea is for profiled type to be a bit of a cleanup, and this seems like a better idea to me.

Krovatkin commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -520,10 +521,10 @@ class ShapePropagator {
             typ->isSubtypeOf(BoolType::get())) {
           node->output()->setType(
-              DimensionedTensorType::create(at::kLong, at::kCPU, 0));
+              ProfiledTensorType::create(at::kLong, at::kCPU, 0, true));
```
ditto

Krovatkin commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -520,10 +521,10 @@ class ShapePropagator {
         } else if (node->input()->type()->isSubtypeOf(FloatType::get())) {
           node->output()->setType(
-              DimensionedTensorType::create(at::kDouble, at::kCPU, 0));
+              ProfiledTensorType::create(at::kDouble, at::kCPU, 0, true));
```
ditto

Krovatkin commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1267,7 +1286,7 @@ class ShapePropagator {
                                         : maybe_dtype_option->toScalarType());
 
-      return {DimensionedTensorType::create(dtype, device, dim)};
+      return {ProfiledTensorType::create(dtype, device, dim, true)};
```
ditto

Krovatkin reviewed (approved):
Looks gut, a few minor nits.

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -204,13 +204,12 @@ std::shared_ptr<FusedKernel> compileKernel(
 
-    // TODO: can't get rid of this use of DimensionedTensorType yet
-    // until we switch to ProfilingGraphExecutor, so we don't have to
-    // run PropagateInputShapes below
```
Well this comment still holds

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -628,25 +672,35 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
         strides_(strides),
         requires_grad_(requires_grad) {}
 
+  ProfiledTensorTypePtr clone() const {
```
Can you elaborate? merge and clone are not the same function.

Krovatkin replied:
Err, I'm sorry, I got the line wrong: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L557 There's `ProfiledTensorType::create` that's equivalent to `clone` as far as I can tell.

zdevito committed:
Update on "Remove DimensionedTensorType"

Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

gh-metadata: pytorch pytorch 24077 gh/zdevito/81/head

zdevito committed:
Update on "Remove DimensionedTensorType"

Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

gh-metadata: pytorch pytorch 24077 gh/zdevito/81/head

zdevito committed:
Update on "Remove DimensionedTensorType"

Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

gh-metadata: pytorch pytorch 24077 gh/zdevito/81/head

zdevito committed:
Update on "Remove DimensionedTensorType"

Remove DimensionedTensorType

This replaces all uses of DimensionedTensorType with ProfiledTensorType.
For places where we propagate shape information, we still follow the
dimension-only propagation rules, meaning that even if full size information
is known on inputs the outputs will only have dimension information.

This fixes several bugs in existing implentations that this change uncovered:
* requires_grad was not propgated correctly across loops
* requires_grad on ProfiledTensorType returned false when requires_grad information
  is unknown but the conservative result is true
* some equality code on ProfiledTensorType contained bugs.

gh-metadata: pytorch pytorch 24077 gh/zdevito/81/head

hunbssfy commented:
> Stack from [ghstack](https://github.com/ezyang/ghstack):
> 
> * **#24077 Remove DimensionedTensorType**
> 
> This replaces all uses of DimensionedTensorType with ProfiledTensorType.
> For places where we propagate shape information, we still follow the
> dimension-only propagation rules, meaning that even if full size information
> is known on inputs the outputs will only have dimension information.
> 
> This fixes several bugs in existing implentations that this change uncovered:
> 
> * requires_grad was not propgated correctly across loops
> * requires_grad on ProfiledTensorType returned false when requires_grad information
>   is unknown but the conservative result is true
> * some equality code on ProfiledTensorType contained bugs.
> 
> This implementation keeps DimensionedTensorType because CompleteTensorType inherits from it. A second patch will remove both, and a third patch will merge ProfiledTensorType with TensorType.
> 
> Differential Revision: [D16729581](https://our.internmc.facebook.com/intern/diff/D16729581)

Hey @zdevio,

May I ask why you only keep the dimension information instead of the full shape information?
I am currently using this pass to do shape inference on a Torchscript graph, I want to get the full shape information for the whole graph based on input shape Infos. Is there any other way to do this if shape_analysis pass won't work?

Thanks!
